kubetest/boskos/common/common.go

```diff
@@ -0,0 +1,301 @@
+limitations under the License.
+*/
+
+package common
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	// Busy state defines a resource being used.
+	Busy = "busy"
+	// Cleaning state defines a resource being cleaned
+	Cleaning = "cleaning"
+	// Dirty state defines a resource that needs cleaning
+	Dirty = "dirty"
+	// Free state defines a resource that is usable
+	Free = "free"
+	// Leased state defines a resource being leased in order to make a new resource
+	Leased = "leased"
+	// ToBeDeleted is used for resources about to be deleted, they will be verified by a cleaner which mark them as tombstone
+	ToBeDeleted = "toBeDeleted"
+	// Tombstone is the state in which a resource can safely be deleted
+	Tombstone = "tombstone"
+	// Other is used to agglomerate unspecified states for metrics reporting
+	Other = "other"
+)
+
+var (
+	// KnownStates is the set of all known states, excluding "other".
+	KnownStates = []string{
+		Busy,
+		Cleaning,
+		Dirty,
+		Free,
+		Leased,
+		ToBeDeleted,
+		Tombstone,
+	}
+)
+
+// UserData is a map of Name to user defined interface, serialized into a string
+type UserData struct {
+	sync.Map
+}
+
+// UserDataMap is the standard Map version of UserMap, it is used to ease UserMap creation.
+type UserDataMap map[string]string
+
+// LeasedResources is a list of resources name that used in order to create another resource by Mason
+type LeasedResources []string
+
+// Duration is a wrapper around time.Duration that parses times in either
+// 'integer number of nanoseconds' or 'duration string' formats and serializes
+// to 'duration string' format.
+type Duration struct {
+	*time.Duration
+}
+
+// UnmarshalJSON implement the JSON Unmarshaler interface in order to be able parse string to time.Duration.
+func (d *Duration) UnmarshalJSON(b []byte) error {
+	if err := json.Unmarshal(b, &d.Duration); err == nil {
+		// b was an integer number of nanoseconds.
+		return nil
+	}
+	// b was not an integer. Assume that it is a duration string.
+
+	var str string
+	err := json.Unmarshal(b, &str)
+	if err != nil {
+		return err
+	}
+
+	pd, err := time.ParseDuration(str)
+	if err != nil {
+		return err
+	}
+	d.Duration = &pd
+	return nil
+}
+
+// Resource abstracts any resource type that can be tracked by boskos
+type Resource struct {
+	Type       string    `json:"type"`
+	Name       string    `json:"name"`
+	State      string    `json:"state"`
+	Owner      string    `json:"owner"`
+	LastUpdate time.Time `json:"lastupdate"`
+	// Customized UserData
+	UserData *UserData `json:"userdata"`
+	// Used to clean up dynamic resources
+	ExpirationDate *time.Time `json:"expiration-date,omitempty"`
+}
+
+// ResourceEntry is resource config format defined from config.yaml
+type ResourceEntry struct {
+	Type     string        `json:"type"`
+	State    string        `json:"state"`
+	Names    []string      `json:"names,flow"`
```

Comment:
I'm looking into the origins of this `flow` statement (which trips up golangci-lint) and it seems to have been there since the very origins of boskos/mason: https://github.com/kubernetes/test-infra/pull/6676/files

It's not a valid `json` string, and I believe we can remove it altogether.. I'm going to add a commit to delete it, but if anyone knows what it's for or why it's needed, please speak now 😄